rules/noUnnecessaryElseRule.ts

```diff
@@ -1,22 +1,56 @@
-import * as ts from 'typescript';
 import * as Lint from 'tslint';
 import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+import * as ts from 'typescript';
 import { isElseIf } from '../src/utils';
 
 const FAIL_MESSAGE = `unnecessary else`;
+const FAIL_MESSAGE_BLOCK = `unnecessary else block`;
+
+const elseBraceRegex = /^[ \t]*else[ \t]\{(.*)\}[ \t]*$/;
+const elseRegex = /^[ \t]*else(.*)$/;
 
 export class Rule extends Lint.Rules.AbstractRule {
-    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
-    }
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
+  }
 }
 
-class IfWalker extends AbstractIfStatementWalker<void> {
-    protected _checkIfStatement(node: ts.IfStatement) {
-        if (node.elseStatement !== undefined &&
-            !isElseIf(node) &&
-            utils.endsControlFlow(node.thenStatement))
-            this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE);
+class IfWalker extends Lint.AbstractWalker<void> {
+  public walk(sourceFile: ts.SourceFile) {
+    const cb = (node: ts.Node): void => {
+      if (node.kind === ts.SyntaxKind.IfStatement && !this._checkIfStatement(<ts.IfStatement>node))
+        return;
+      return ts.forEachChild(node, cb);
+    };
+
+    return ts.forEachChild(sourceFile, cb);
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    const elseStatement = node.elseStatement;
+
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      // remove else scope if safe: keep blocks where local variables change scope when unwrapped
+      if (isElseStatement(elseStatement) && utils.isBlock(elseStatement) && !hasLocals(elseStatement)) {
+        const replacement = elseStatement.getText().replace(elseBraceRegex, '$1').replace(elseRegex, '$1');
```

Comment:
I commented on that earlier, so here's a more detailed explanation:
The only reason you need to abort checking child nodes is because you replace the whole statement (fixes for nested nodes are then inside this already fixed range) and TSLint really sucks at handling overlapping fixes.
Actually you only want to remove `else {` and `}`. Here's how I would approach that:

```ts
const elseKeyword = node.getChildAt(5 /*else*/, this.sourceFile);
const fixBlock = [
  Lint.Replacement.deleteFromTo(elseKeyword.end - 4, elseStatement.statements.pos), // removes `else {`
  Lint.Replacement.deleteText(elseStatement.end - 1, 1), // deletes `}`
]
```